File: src/petridish/schema.py
"""The frozen state schema for ``~/.petridish/projects.json``.

This module is the shared contract: the daemon writes it, every frontend reads
it, and each sensor produces :class:`AgentSignal` values that the aggregator
folds into :class:`Project` records.  See ``IMPLEMENTATION_PLAN.md`` §4.

Two rules matter more than the rest:

* **Timestamps are second-resolution UTC**, serialised as ``...Z`` strings.
  Sub-second precision is deliberately discarded — nothing in this system
  reasons about milliseconds, and truncating keeps round-tripping exact.
* **Writes are atomic** (see :func:`write_atomic`).  The daemon is the sole
  writer; readers therefore never observe a partial file and need no lock.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field, replace
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Mapping, TypedDict

SCHEMA_VERSION = 1

#: `~/.petridish/` — the daemon's config/state directory. Shared across every
#: frontend (the `petri` TUI, `menubar.py`, `installer.py`) and the Rust
#: scanner (`swab-rs/`), which resolves the identical path independently on
#: its own side (`config::default_path()`).
CONFIG_DIR = os.path.join(os.path.expanduser("~"), ".petridish")

#: Default `projects.json` state path — the file the Rust scanner
#: (`swab scan`) writes and every Python frontend reads via `read_json`.
_DEFAULT_STATE_PATH = os.path.join(CONFIG_DIR, "projects.json")

#: Structural marker tagging every hook entry `installer.py` writes into
#: `~/.claude/settings.json`, so `--uninstall` can remove exactly (and only)
#: what this project added without disturbing other hook consumers.
HOOK_MARKER = "# petridish"

# ---------------------------------------------------------------------------
# Serialised shapes.
#
# These TypedDicts are the wire contract for ``~/.petridish/projects.json`` —
# the same shape Raycast's `raycast/src/types.ts` mirrors by hand. Declaring
# them here makes that contract machine-checked: add a field to a dataclass and
# forget it in ``to_dict``, and pyright fails instead of a frontend silently
# reading ``undefined``.
#
# Note the deliberate asymmetry with ``from_dict``. ``to_dict`` returns a
# TypedDict because *we* produce that shape and guarantee it. ``from_dict``
# takes a permissive ``Mapping[str, Any]`` because it parses whatever is
# actually on disk — a hand-edited, truncated, or older-version file is
# expected input, not a type error. Typing the input strictly would be a lie
# that pushes real parsing failures to runtime.
# ---------------------------------------------------------------------------


class GitStateDict(TypedDict):
    is_repo: bool
    branch: str | None
    is_dirty: bool
    uncommitted_files: int
    last_commit_at: str | None
    mine_last_commit_at: str | None
    github_url: str | None


class AgentStateDict(TypedDict):
    state: str
    active_agent: str | None
    last_event: str | None
    last_event_at: str | None
    session_id: str | None


# Allowed enum-ish values. Kept as plain tuples rather than Enums so the JSON
# stays trivially readable from TypeScript/jq without a mapping layer.
AGENT_STATES = ("working", "recent", "idle")
STATUS_BUCKETS = ("active", "in_flight", "stale", "cold")

#: Silence thresholds, in seconds, that define :data:`AGENT_STATES`.
#:
#: These live here rather than in ``scan.py`` because two clocks read them. The
#: daemon stamps ``agent.state`` at *scan* time, but a frontend showing a live
#: "silent 4m 12s" counter must re-derive the state at *render* time — a
#: ``projects.json`` that is 90 seconds old already has a stale ``state`` field.
#: With the numbers duplicated, the traffic light and the clock beside it would
#: eventually disagree.
AGENT_WORKING_MAX_S = 90
AGENT_RECENT_MAX_S = 30 * 60


def agent_state_for_silence(silence_s: float) -> str:
    """Map seconds-since-last-event onto one of :data:`AGENT_STATES`.

    The sole definition of what the three states *mean*. Negative input (a
    clock skew between the writer and the reader) is treated as zero rather
    than raising — a frontend must never crash on a timestamp from the future.
    """
    if silence_s < 0:
        silence_s = 0.0
    if silence_s < AGENT_WORKING_MAX_S:
        return "working"
    if silence_s < AGENT_RECENT_MAX_S:
        return "recent"
    return "idle"


# ---------------------------------------------------------------------------
# datetime helpers
# ---------------------------------------------------------------------------

def to_utc(dt: datetime) -> datetime:
    """Normalise to tz-aware UTC at second resolution.

    A naive datetime is *assumed* to be UTC rather than local time — the
    daemon only ever deals in UTC, and guessing local time here would silently
    shift every timestamp by the machine's offset.
    """
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc).replace(microsecond=0)


def utcnow() -> datetime:
    """Current time, tz-aware UTC, second resolution."""
    return datetime.now(timezone.utc).replace(microsecond=0)


def _iso(dt: datetime | None) -> str | None:
    """Serialise to ``2026-08-05T22:45:00Z``; ``None`` passes through."""
    if dt is None:
        return None
    return to_utc(dt).strftime("%Y-%m-%dT%H:%M:%SZ")


def _parse(s: str | None) -> datetime | None:
    """Inverse of :func:`_iso`. Returns a tz-aware UTC datetime."""
    if s is None:
        return None
    # fromisoformat handles "+00:00" natively; normalise the "Z" suffix first.
    return to_utc(datetime.fromisoformat(s.replace("Z", "+00:00")))


# ---------------------------------------------------------------------------
# Records
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class GitState:
    """Git facts for one project. ``is_repo=False`` is the safe degraded value."""

    is_repo: bool = False
    branch: str | None = None
    is_dirty: bool = False
    uncommitted_files: int = 0
    last_commit_at: datetime | None = None
    #: Last commit authored by *this user* (see the authorship filter, plan §2).
    #: Preferred over ``last_commit_at`` when bucketing, so a teammate's or a
    #: bot's push doesn't make a project look like yours.
    mine_last_commit_at: datetime | None = None
    github_url: str | None = None

    def to_dict(self) -> GitStateDict:
        return {
            "is_repo": self.is_repo,
            "branch": self.branch,
            "is_dirty": self.is_dirty,
            "uncommitted_files": self.uncommitted_files,
            "last_commit_at": _iso(self.last_commit_at),
            "mine_last_commit_at": _iso(self.mine_last_commit_at),
            "github_url": self.github_url,
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> GitState:
        return cls(
            is_repo=d.get("is_repo", False),
            branch=d.get("branch"),
            is_dirty=d.get("is_dirty", False),
            uncommitted_files=d.get("uncommitted_files", 0),
            last_commit_at=_parse(d.get("last_commit_at")),
            mine_last_commit_at=_parse(d.get("mine_last_commit_at")),
            github_url=d.get("github_url"),
        )


@dataclass(frozen=True)
class AgentState:
    """Live agent activity for one project, as published to frontends."""

    state: str = "idle"
    active_agent: str | None = None
    last_event: str | None = None
    last_event_at: datetime | None = None
    session_id: str | None = None

    def to_dict(self) -> AgentStateDict:
        return {
            "state": self.state,
            "active_agent": self.active_agent,
            "last_event": self.last_event,
            "last_event_at": _iso(self.last_event_at),
            "session_id": self.session_id,
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> AgentState:
        return cls(
            state=d.get("state", "idle"),
            active_agent=d.get("active_agent"),
            last_event=d.get("last_event"),
            last_event_at=_parse(d.get("last_event_at")),
            session_id=d.get("session_id"),
        )


@dataclass(frozen=True)
class AgentSignal:
    """One sensor's observation. **Internal — never serialised into projects.json.**

    Every sensor (M4 Claude, M5 Copilot, M6 hook events) returns
    ``dict[str, AgentSignal]`` keyed by ``root``, already collapsed to the
    newest signal per root.  The aggregator merges across sensors by the same
    newest-wins rule.

    ``root`` is a *resolved project root* — the sensor has already walked it up
    from the raw cwd via ``discovery.resolve_root`` — so that a monorepo session
    in ``repo/packages/core`` attributes to ``repo`` rather than fragmenting
    into phantom projects.  ``raw_cwd`` keeps the pre-resolution value for
    debugging that attribution.
    """

    root: str
    at: datetime
    agent: str
    session_id: str | None = None
    event: str | None = None
    raw_cwd: str | None = None


class QuotaStateDict(TypedDict):
    measured_at: str | None
    five_hour_used_pct: int | None
    five_hour_resets_at: str | None
    seven_day_used_pct: int | None
    seven_day_resets_at: str | None
    context_used_pct: int | None


@dataclass(frozen=True)
class QuotaState:
    """Claude subscription usage, as last reported by Claude Code itself.

    Sourced from ``~/.claude/last-status.json`` — see `swab-rs`'s
    ``sensors::quota`` module. Account-global, not per-project: this
    belongs in a header, never in a project row.

    Every field is optional and defaults to ``None``. That is not defensive
    padding: the file is an **undocumented internal** of another program, so
    any field may disappear on a Claude Code upgrade, and a missing value must
    degrade to "unknown" rather than break the tick (invariant 5).

    ``measured_at`` is when *Claude Code* wrote the numbers, not when the
    daemon read them. A frontend needs it to say how stale the figures are:
    the file only updates while a session is running, so overnight it can be
    hours old while everything else in ``projects.json`` is a minute old.
    """

    measured_at: datetime | None = None
    five_hour_used_pct: int | None = None
    five_hour_resets_at: datetime | None = None
    seven_day_used_pct: int | None = None
    seven_day_resets_at: datetime | None = None
    context_used_pct: int | None = None

    def to_dict(self) -> QuotaStateDict:
        return {
            "measured_at": _iso(self.measured_at),
            "five_hour_used_pct": self.five_hour_used_pct,
            "five_hour_resets_at": _iso(self.five_hour_resets_at),
            "seven_day_used_pct": self.seven_day_used_pct,
            "seven_day_resets_at": _iso(self.seven_day_resets_at),
            "context_used_pct": self.context_used_pct,
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> QuotaState:
        return cls(
            measured_at=_parse(d.get("measured_at")),
            five_hour_used_pct=d.get("five_hour_used_pct"),
            five_hour_resets_at=_parse(d.get("five_hour_resets_at")),
            seven_day_used_pct=d.get("seven_day_used_pct"),
            seven_day_resets_at=_parse(d.get("seven_day_resets_at")),
            context_used_pct=d.get("context_used_pct"),
        )


class ProjectDict(TypedDict):
    id: str
    name: str
    path: str
    category: str
    is_foreign: bool
    git: GitStateDict
    agent: AgentStateDict
    last_activity_at: str | None
    status_bucket: str


@dataclass(frozen=True)
class Project:
    """One tracked project."""

    id: str
    name: str
    path: str
    category: str
    is_foreign: bool = False
    git: GitState = field(default_factory=GitState)
    agent: AgentState = field(default_factory=AgentState)
    last_activity_at: datetime | None = None
    status_bucket: str = "cold"

    def to_dict(self) -> ProjectDict:
        return {
            "id": self.id,
            "name": self.name,
            "path": self.path,
            "category": self.category,
            "is_foreign": self.is_foreign,
            "git": self.git.to_dict(),
            "agent": self.agent.to_dict(),
            "last_activity_at": _iso(self.last_activity_at),
            "status_bucket": self.status_bucket,
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> Project:
        return cls(
            id=d["id"],
            name=d["name"],
            path=d["path"],
            category=d["category"],
            is_foreign=d.get("is_foreign", False),
            git=GitState.from_dict(d.get("git") or {}),
            agent=AgentState.from_dict(d.get("agent") or {}),
            last_activity_at=_parse(d.get("last_activity_at")),
            status_bucket=d.get("status_bucket", "cold"),
        )


class RadarDict(TypedDict):
    schema_version: int
    updated_at: str | None
    scan_duration_ms: int
    projects: list[ProjectDict]
    #: Always present (possibly ``null``) so the shape we *write* is total,
    #: even though what we *read* may predate the field entirely.
    quota: QuotaStateDict | None


@dataclass(frozen=True)
class Radar:
    """The whole ``projects.json`` document."""

    updated_at: datetime
    projects: tuple[Project, ...] = ()
    scan_duration_ms: int = 0
    schema_version: int = SCHEMA_VERSION
    #: Account-wide Claude quota, or ``None`` when the sensor found nothing.
    #: Optional and last so every existing construction site keeps working.
    quota: QuotaState | None = None

    def to_dict(self) -> RadarDict:
        return {
            "schema_version": self.schema_version,
            "updated_at": _iso(self.updated_at),
            "scan_duration_ms": self.scan_duration_ms,
            "projects": [p.to_dict() for p in self.projects],
            "quota": self.quota.to_dict() if self.quota else None,
        }

    @classmethod
    def from_dict(cls, d: Mapping[str, Any]) -> Radar:
        updated = _parse(d.get("updated_at"))
        if updated is None:
            raise ValueError("projects.json is missing required 'updated_at'")
        return cls(
            updated_at=updated,
            # Must be a tuple: a list would break both frozen-ness and the
            # round-trip equality contract.
            projects=tuple(Project.from_dict(p) for p in d.get("projects", ())),
            scan_duration_ms=d.get("scan_duration_ms", 0),
            schema_version=d.get("schema_version", SCHEMA_VERSION),
            # Absent in every file written before this field existed, and null
            # whenever the sensor came up empty. Both mean "unknown", not "0%".
            quota=(
                QuotaState.from_dict(d["quota"])
                if isinstance(d.get("quota"), dict)
                else None
            ),
        )

    def to_json(self, *, indent: int | None = 2) -> str:
        return json.dumps(self.to_dict(), indent=indent) + "\n"

    @classmethod
    def from_json(cls, text: str) -> Radar:
        return cls.from_dict(json.loads(text))


# ---------------------------------------------------------------------------
# Atomic write
# ---------------------------------------------------------------------------

def write_atomic(radar: Radar, path: str | os.PathLike) -> None:
    """Write ``radar`` to ``path`` atomically.

    Serialises to a sibling ``.tmp`` file, then :func:`os.replace` onto the
    target — ``os.replace`` is atomic within a filesystem, so a concurrent
    reader sees either the old file or the new one, never a half-written one.
    The temp file must be a *sibling* (not in /tmp) for that guarantee to hold.

    The parent directory is created if missing.  On any failure the temp file is
    removed rather than left behind for the next run to trip over.
    """
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    tmp = target.with_name(target.name + ".tmp")
    try:
        tmp.write_text(radar.to_json(), encoding="utf-8")
        os.replace(tmp, target)
    except BaseException:
        try:
            tmp.unlink(missing_ok=True)
        except OSError:
            pass
        raise


def read_json(path: str | os.PathLike) -> Radar:
    """Load a :class:`Radar` from disk. Raises ``FileNotFoundError`` if absent."""
    return Radar.from_json(Path(path).read_text(encoding="utf-8"))


__all__ = [
    "SCHEMA_VERSION",
    "AGENT_STATES",
    "STATUS_BUCKETS",
    "AGENT_WORKING_MAX_S",
    "AGENT_RECENT_MAX_S",
    "agent_state_for_silence",
    "GitState",
    "AgentState",
    "QuotaState",
    "AgentSignal",
    "Project",
    "Radar",
    "write_atomic",
    "read_json",
    "to_utc",
    "utcnow",
    "replace",
]

File: petridish-core/src/schema.rs
//! Wire contract for `projects.json`. Mirrors `src/petridish/schema.py` field-for-field.
//! Field types/names here are the scaffold contract — module R1 fills in the logic
//! (`write_atomic`, `agent_state_for_silence`, bucketing) without changing these shapes.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::Path;

/// Serde helpers truncating every `DateTime<Utc>` to whole-second precision at the
/// serialization boundary, matching the real wire contract (`petridish.schema.to_utc`/
/// `_iso`, which apply `.replace(microsecond=0)` regardless of where the timestamp
/// originated -- mtime reads, git commit times, etc. can all carry sub-second precision
/// in memory). Without this, a value round-tripped straight from `serde`'s default
/// `DateTime<Utc>` impl (full nanosecond precision) diverges from the Python output on
/// every mtime-derived field (`AgentState.last_event_at`, `Project.last_activity_at`) --
/// caught only once the full aggregator pipeline existed end to end and `diff_check.sh`
/// could run for the first time (see the R9 commit message for how this surfaced).
mod iso_second {
    use chrono::{DateTime, SubsecRound, Utc};
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S: Serializer>(dt: &DateTime<Utc>, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&dt.trunc_subsecs(0).format("%Y-%m-%dT%H:%M:%SZ").to_string())
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(d: D) -> Result<DateTime<Utc>, D::Error> {
        let s = String::deserialize(d)?;
        DateTime::parse_from_rfc3339(&s)
            .map(|dt| dt.with_timezone(&Utc))
            .map_err(serde::de::Error::custom)
    }
}

mod iso_second_opt {
    use chrono::{DateTime, SubsecRound, Utc};
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S: Serializer>(
        dt: &Option<DateTime<Utc>>,
        s: S,
    ) -> Result<S::Ok, S::Error> {
        match dt {
            Some(dt) => {
                s.serialize_str(&dt.trunc_subsecs(0).format("%Y-%m-%dT%H:%M:%SZ").to_string())
            }
            None => s.serialize_none(),
        }
    }

    pub fn deserialize<'de, D: Deserializer<'de>>(
        d: D,
    ) -> Result<Option<DateTime<Utc>>, D::Error> {
        let opt: Option<String> = Option::deserialize(d)?;
        match opt {
            Some(s) => DateTime::parse_from_rfc3339(&s)
                .map(|dt| Some(dt.with_timezone(&Utc)))
                .map_err(serde::de::Error::custom),
            None => Ok(None),
        }
    }
}

/// Silence below this many seconds => `AgentActivity::Working`.
pub const AGENT_WORKING_MAX_S: i64 = 90;
/// Silence below this many seconds (and not `Working`) => `AgentActivity::Recent`.
pub const AGENT_RECENT_MAX_S: i64 = 1800;

/// Samples kept in `Project::agent_activity`, one appended per `swab scan` tick (~60s tick
/// interval -> the full ring spans roughly one hour). Carried forward across ticks by the
/// aggregator itself -- `events.ndjson` cannot supply this history on its own, since it's
/// compacted to a single newest-signal-per-root and truncated every tick (see
/// `swab/src/events.rs`'s module doc). Investigated and confirmed via a real read of this
/// machine's `~/.petridish/events.ndjson` before this field was added.
pub const AGENT_ACTIVITY_WINDOW: usize = 60;

/// Trailing days of daily commit counts kept in `GitState::daily_commits`. Unlike agent
/// activity, this needs no cross-tick carry-forward -- git already retains its own commit
/// history, so it's recomputed fresh from real history every tick.
pub const GIT_ACTIVITY_WINDOW_DAYS: usize = 14;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentActivity {
    Working,
    Recent,
    Idle,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StatusBucket {
    Active,
    InFlight,
    Stale,
    Cold,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GitState {
    pub is_repo: bool,
    pub branch: Option<String>,
    pub is_dirty: bool,
    pub uncommitted_files: u32,
    #[serde(with = "iso_second_opt")]
    pub last_commit_at: Option<DateTime<Utc>>,
    #[serde(with = "iso_second_opt")]
    pub mine_last_commit_at: Option<DateTime<Utc>>,
    pub github_url: Option<String>,
    /// Daily commit counts for the trailing `GIT_ACTIVITY_WINDOW_DAYS` days, oldest first,
    /// today last. Empty when `is_repo` is false. `#[serde(default)]` so a `projects.json`
    /// written before this field existed still deserializes.
    #[serde(default)]
    pub daily_commits: Vec<u32>,
}

impl GitState {
    /// `GitState { is_repo: false, .. }` — the required "never an exception" fallback
    /// for a failed/timed-out git call (invariant #6).
    pub fn not_a_repo() -> Self {
        GitState {
            is_repo: false,
            branch: None,
            is_dirty: false,
            uncommitted_files: 0,
            last_commit_at: None,
            mine_last_commit_at: None,
            github_url: None,
            daily_commits: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentState {
    pub state: AgentActivity,
    pub active_agent: Option<String>,
    pub last_event: Option<String>,
    #[serde(with = "iso_second_opt")]
    pub last_event_at: Option<DateTime<Utc>>,
    pub session_id: Option<String>,
}

impl AgentState {
    /// The degrade-never-abort default (invariant #5) — no signal for this project.
    pub fn idle_unknown() -> Self {
        AgentState {
            state: AgentActivity::Idle,
            active_agent: None,
            last_event: None,
            last_event_at: None,
            session_id: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Project {
    pub id: String,
    pub name: String,
    pub path: String,
    pub category: String,
    #[serde(default)]
    pub parent_path: Option<String>,
    pub is_foreign: bool,
    pub git: GitState,
    pub agent: AgentState,
    #[serde(with = "iso_second_opt")]
    pub last_activity_at: Option<DateTime<Utc>>,
    pub status_bucket: StatusBucket,
    /// Per-tick agent-event counts for the trailing `AGENT_ACTIVITY_WINDOW` ticks, oldest
    /// first, this tick last. Maintained by `swab scan` carrying it forward from the previous
    /// `projects.json` and appending one new sample per tick -- see `AGENT_ACTIVITY_WINDOW`'s
    /// doc for why this can't be derived from `events.ndjson` alone. `#[serde(default)]` so a
    /// `projects.json` written before this field existed still deserializes.
    #[serde(default)]
    pub agent_activity: Vec<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct QuotaState {
    #[serde(with = "iso_second_opt")]
    pub measured_at: Option<DateTime<Utc>>,
    pub five_hour_used_pct: Option<u8>,
    #[serde(with = "iso_second_opt")]
    pub five_hour_resets_at: Option<DateTime<Utc>>,
    pub seven_day_used_pct: Option<u8>,
    #[serde(with = "iso_second_opt")]
    pub seven_day_resets_at: Option<DateTime<Utc>>,
    pub context_used_pct: Option<u8>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Radar {
    pub schema_version: u32,
    #[serde(with = "iso_second")]
    pub updated_at: DateTime<Utc>,
    pub scan_duration_ms: u64,
    pub projects: Vec<Project>,
    pub quota: Option<QuotaState>,
}

/// Internal sensor contract — never serialized into `projects.json` (`Radar`/`Project`
/// never hold one). `Serialize` here is for `examples/probe.rs`'s parity-check tooling only.
/// `root` is always a `resolve_root()`-resolved path (invariant #3), never a raw `cwd`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct AgentSignal {
    pub root: String,
    #[serde(with = "iso_second")]
    pub at: DateTime<Utc>,
    pub agent: String,
    pub session_id: Option<String>,
    pub event: Option<String>,
    pub raw_cwd: Option<String>,
}

/// Silence (seconds since last event) -> `AgentActivity`, per `AGENT_WORKING_MAX_S` /
/// `AGENT_RECENT_MAX_S`. Negative silence must clamp to zero, not panic or go negative.
///
/// Mirrors the Python reference (`petridish.schema.agent_state_for_silence`):
/// negative input is treated as zero rather than panicking, a frontend must
/// never crash on a timestamp from the future.
pub fn agent_state_for_silence(silence_seconds: i64) -> AgentActivity {
    let silence = if silence_seconds < 0 { 0 } else { silence_seconds };
    if silence < AGENT_WORKING_MAX_S {
        AgentActivity::Working
    } else if silence < AGENT_RECENT_MAX_S {
        AgentActivity::Recent
    } else {
        AgentActivity::Idle
    }
}

/// Serialize `radar` to `<path>.tmp` (same dir as `path`) then atomically rename onto
/// `path` (invariant #1: daemon is the sole writer, temp-file + rename). On any failure
/// the tmp file must be removed; the parent dir is created if missing.
pub fn write_atomic(path: &Path, radar: &Radar) -> std::io::Result<()> {
    // Same behavior as `petridish.schema.write_atomic`: parent dir first,
    // sibling `.tmp` for the rename to stay on the same filesystem (so
    // `os.replace` / `fs::rename` stays atomic), and cleanup on any failure.
    std::fs::create_dir_all(path.parent().ok_or_else(|| {
        std::io::Error::new(
            std::io::ErrorKind::Other,
            "radar path has no parent directory",
        )
    })?)?;

    let tmp_name = path
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "radar path has a non-UTF8 filename",
            )
        })?;
    let tmp = path.with_file_name(format!("{}.tmp", tmp_name));

    let body = serde_json::to_string(radar).map_err(|e| {
        std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
    })?;

    std::fs::write(&tmp, &body)?;
    std::fs::rename(&tmp, path)
        .map_err(|rename_err| {
            // Best-effort cleanup: ignore any error from this remove so the
            // original rename failure is what the caller sees.
            let _ = std::fs::remove_file(&tmp);
            rename_err
        })
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::SubsecRound;

    #[test]
    fn agent_state_for_silence_zero_is_working() {
        assert_eq!(agent_state_for_silence(0), AgentActivity::Working);
    }

    #[test]
    fn agent_state_for_silence_just_under_working_is_working() {
        assert_eq!(agent_state_for_silence(89), AgentActivity::Working);
    }

    #[test]
    fn agent_state_for_silence_working_upper_bound_exclusive() {
        // AGENT_WORKING_MAX_S (90) is *not* Working — boundary is exclusive.
        assert_eq!(agent_state_for_silence(90), AgentActivity::Recent);
    }

    #[test]
    fn agent_state_for_silence_just_under_recent_is_recent() {
        assert_eq!(agent_state_for_silence(1799), AgentActivity::Recent);
    }

    #[test]
    fn agent_state_for_silence_recent_upper_bound_is_idle() {
        // AGENT_RECENT_MAX_S (1800) is *not* Recent — falls through to Idle.
        assert_eq!(agent_state_for_silence(1800), AgentActivity::Idle);
    }

    #[test]
    fn agent_state_for_silence_negative_clamps_to_zero() {
        // Negative input must clamp to 0 -> Working, never panic.
        assert_eq!(agent_state_for_silence(-50), AgentActivity::Working);
    }

    #[test]
    fn write_atomic_creates_missing_parent_dir_and_no_tmp_left_behind() {
        let tmp = std::env::temp_dir();
        let dir = tmp.join(format!("swab_write_atomic_test_dir_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir); // start clean

        let path = dir.join("projects.json");
        assert!(!path.exists());
        assert!(!path.parent().unwrap().exists());

        let radar = Radar {
            schema_version: 1,
            updated_at: chrono::Utc::now().trunc_subsecs(0),
            scan_duration_ms: 0,
            projects: vec![],
            quota: None,
        };
        write_atomic(&path, &radar).expect("write_atomic should succeed");

        assert!(path.exists(), "target file must exist after write_atomic");
        // No stray .tmp sibling should be left behind.
        assert!(
            !path.with_file_name("projects.json.tmp").exists(),
            "stray .tmp file must not be left behind"
        );

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn write_atomic_overwrite_second_call_differs() {
        let tmp = std::env::temp_dir();
        let dir = tmp.join(format!("swab_write_atomic_overwrite_test_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);

        let path = dir.join("projects.json");
        let radar_a = Radar {
            schema_version: 1,
            updated_at: chrono::Utc::now().trunc_subsecs(0),
            scan_duration_ms: 0,
            projects: vec![],
            quota: None,
        };
        write_atomic(&path, &radar_a).expect("first write should succeed");

        // Second call: use a different `scan_duration_ms` so the contents differ.
        let mut radar_b = radar_a.clone();
        radar_b.scan_duration_ms = 9999;
        write_atomic(&path, &radar_b).expect("second write should succeed");

        // Read back the file on disk.
        let back = std::fs::read_to_string(&path).expect("file must be readable");

        // File contents must match radar_b, not radar_a.
        let file_json: serde_json::Value =
            serde_json::from_str(&back).expect("file is valid JSON");

        let expected_b: serde_json::Value =
            serde_json::from_str(&serde_json::to_string(&radar_b).unwrap()).unwrap();
        assert_eq!(file_json, expected_b);
    }

    #[test]
    fn write_atomic_round_trip_minimal_radar() {
        let radar = Radar {
            schema_version: 1,
            updated_at: chrono::Utc::now().trunc_subsecs(0),
            scan_duration_ms: 0,
            projects: vec![],
            quota: None,
        };
        let s = serde_json::to_string(&radar).expect("serialize minimal Radar");
        let decoded: Radar = serde_json::from_str(&s).expect("round-trip minimal Radar");
        assert_eq!(radar, decoded);
    }
}
